### PR TITLE
Add mm_pattern and bmm_pattern to serialized_patterns

### PR DESCRIPTION
**Pull request by aorenste**
Make it easier to serialize patterns by adding `pattern_matcher.gen_register_replacement()` which is like `pattern_matcher.register_replacement()` but also requires the replacement to be precompiled.

To precompile patterns (and save to disk) run:
```
torchgen/fuse_attention_patterns/gen_attention_patterns.py
```

- Updated the sfdp patterns to use `gen_register_replacement`.
- Add serialized patterns for mm_pattern and bmm_pattern (The 'misc' patterns don't serialize cleanly so can't be added).
- Updated the testing so it checked the round-trip patterns match and not just that it serialized the same way.
- Checking that the patterns round-trip properly found that the `users` field wasn't being serialized properly.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #121314
* __->__ #121313

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Review by eellison — COMMENTED** [review]
Cool! just curious, any exact compilation time increase from this ?

Would it be possible to add syntax so that it is easier to use serialization for future patterns? I'm imagining something like
```
register_replacement(
    gen_pattern_fn(fn, unique_key
```

which when run with an environment variable would serialize the pattern for you. And then subsequently would load the previously serialized pattern. Basically, the same UX as pytorch expect tests. 

cc @Chillee who i previously talked to abt this

**Comment by aorenste** [comment]
> Cool! just curious, any exact compilation time increase from this ?

The compilation time improvement actually comes from the next PR in the stack (#121314).  I'm waiting for benchmarks to run so I can measure it.

> Would it be possible to add syntax so that it is easier to use serialization for future patterns? I'm imagining something like
> 
> ```
> register_replacement(
>     gen_pattern_fn(fn, unique_key
> ```
> 
> which when run with an environment variable would serialize the pattern for you. And then subsequently would load the previously serialized pattern. Basically, the same UX as pytorch expect tests.

Neat idea - I'll try that.

**Comment by aorenste** [comment]
> >   
> > which when run with an environment variable would serialize the pattern for you. And then subsequently would load the previously serialized pattern. Basically, the same UX as pytorch expect tests.
> 
> Neat idea - I'll try that.

@eellison So I thought about this a bit more. Unless I'm missing the point of what you're saying I'm not sure it can really get much simpler than it is. The only real difference is that you run a script (gen_attention_patterns.py) instead of setting an environment variable. I'm not sure there's a lot of value in baking the contents of that script into the pytorch runtime.

Since we still need a common registry of some sort you still need to tell *something* global about which generators to probe.

**Comment by eellison** [comment]
> The only real difference is that you run a script (gen_attention_patterns.py) instead of setting an environment variable

The difference is there's much code changes required in order to get it to work.

> Since we still need a common registry of some sort you still need to tell something global about which generators to probe

I'm not sure we need a global registry, just the generation of particular pattern needs to know if a serialized version exists. Which would be the same lookup that writes the pattern if the env variable was set.

I'm still okay landing without it but I do think it would be an improvement (and an easier sell in adoption for internal/external teams with patterns).

In either case could we add some tests along these lines: https://github.com/pytorch/pytorch/blob/main/test/inductor/test_pattern_matcher.py#L1004-L1008?

**Comment by aorenste** [comment]
@eellison Ok - updated to a form that I think is closer to what you suggested.

Also added full testing - which ended up catching a bug in the original sfdp patterns (the `users` parameter wasn't being written when it was `0`).

**Comment by eellison** [comment]
cool - will take a look when test failures are fixed

File: torch/_inductor/constant_folding.py [hunk 1]
```diff
@@ -102,7 +103,8 @@ def run_node(self, node):
             def set_env(arg):
                 self.env[arg] = self.unknown_value
 
-            pytree.tree_map_only(torch.fx.Node, set_env, node.args)
+            # In-place is fine since we don't mutate
+            pytree.tree_map_only_(torch.fx.Node, set_env, node.args)
```
Comment (eellison) [inline torch/_inductor/constant_folding.py:107]: nit: unrelated..

File: torch/_inductor/fx_passes/serialized_patterns/_sfdp_pattern_1.py [hunk 2]
```diff
@@ -100,7 +100,7 @@
 expand_default_3 = CallFunction(aten.expand.default, KeywordArg('value'), Ignored())
 view_default_4 = CallFunction(aten.view.default, expand_default_3, Ignored())
 bmm_default_1 = CallFunction(aten.bmm.default, view_default_3, view_default_4)
-_sfdp_pattern_1_inference = CallFunction(aten.view.default, bmm_default_1, Ignored())
+_sfdp_pattern_1_inference = CallFunction(aten.view.default, bmm_default_1, Ignored(), _users=0)
```
Comment (eellison) [inline torch/_inductor/fx_passes/serialized_patterns/_sfdp_pattern_1.py:103]: Is there a reason to add `_user=0` here ?

File: torch/_inductor/fx_passes/serialized_patterns/addmm_pattern.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
+# mypy: ignore-errors
+
+# noqa: F401, E501
+# This is an auto-generated file. Please do not modify it by hand.
+# To re-generate, run:
+# cd ~/pytorch && python
+# torchgen/fuse_attention_patterns/gen_attention_patterns.py
```
Comment (eellison) [inline torch/_inductor/fx_passes/serialized_patterns/addmm_pattern.py:7]: Not relevant here right ?

File: torch/_inductor/pattern_matcher.py [hunk 4]
```diff
@@ -229,6 +236,14 @@ def find_anchor_nodes(self, ctx: MatchContext, searched):
         if self in ctx.pattern_to_node:
             yield ctx.pattern_to_node[self]
 
+    def pattern_eq(self, other: Any) -> bool:
```
Comment (eellison) [inline torch/_inductor/pattern_matcher.py:239]: Is there a reason to define `pattern_eq` instead of `__eq__` ?

File: torch/_inductor/fx_passes/misc_patterns.py [hunk 5]
```diff
@@ -58,7 +58,7 @@ def randperm_index_replacement(x, slice_shape):
         index = torch.randperm(x.shape[0], device=x.device)[:slice_shape]
         return torch.ops.aten._unsafe_index(x, (index,)), index
 
-    pattern = register_replacement(
+    register_replacement(
```
Comment (eellison) [inline torch/_inductor/fx_passes/misc_patterns.py:61]: Doesn't look like we serialized this one

**Review by eellison — COMMENTED** [review]
Looks great! A few comments. Also curious @Chillee's thoughts.

Might be easier to land just with addmm & others first and then update the fuse attention. In one go is fine too.

File: torch/_inductor/constant_folding.py (hunk 1 above)
Comment (aorenste) [reply torch/_inductor/constant_folding.py:107]: True (Related only in that I discovered it while looking at the perf of the pattern matcher). I kind of figured for a single-line fix it wasn't worth a PR on its own (especially considering how long it takes to test/review a PR).

File: torch/_inductor/fx_passes/serialized_patterns/_sfdp_pattern_1.py (hunk 2 above)
Comment (aorenste) [reply torch/_inductor/fx_passes/serialized_patterns/_sfdp_pattern_1.py:103]: As mentioned in the summary - the patterns weren't being serialized properly before and the test didn't catch it. This is the output of the generator now that it's correctly serializing the pattern.

File: torch/_inductor/fx_passes/serialized_patterns/addmm_pattern.py (hunk 3 above)
Comment (aorenste) [reply torch/_inductor/fx_passes/serialized_patterns/addmm_pattern.py:7]: Which part? It's auto-generated and you can use that script to regenerate it (the script is poorly named now since it regenerates ALL patterns, not just fuse_attention_patterns - maybe that's your issue?)

File: torch/_inductor/pattern_matcher.py (hunk 4 above)
Comment (aorenste) [reply torch/_inductor/pattern_matcher.py:239]: I originally defined `__eq__`.  The main reason I changed it is I didn't really want to have to implement `__hash__` since I only care about this for debugging.

File: test/inductor/test_pattern_matcher.py [hunk 6]
```diff
@@ -947,66 +943,50 @@ def fn2(inp, a, b):
...
                     PatternPrettyPrinter.run(search_fn_pattern),
-                    msg=f"Found mismatched pattern {key}. Run gen_attention_patterns.py",
+                    msg=f"Found mismatched pattern {search_fn.__name__}. Run gen_attention_patterns.py",
                 )
 
+                # Since we've already checked that the serialized patterns match
+                # lets verify the serializer by ensuring the generated patterns
+                # also match (since search_fn_pattern is the serialized version
+                # of search_fn).
+                self.assertTrue(pattern.pattern_eq(search_fn_pattern))
```
[the remark this answers is not in the document]
Comment (aorenste) [reply test/inductor/test_pattern_matcher.py:988]: Not sufficient (I found a serialization bug with `_users` by doing it this way).

The problem is if you just check the output then you're just checking that the transformation to the serialized form is consistent - not that you actually serialized the whole thing:
Imagine serializing a tuple `(A, B, C)` where the serializer dropped `B`:
- you'd serialize as `(A, C)`
- read that in as `(A, C)` 
- and then write it back out as `(A, C)`.
The serialized output is the same but it's clearly dropping something.

File: torch/_inductor/pattern_matcher.py [hunk 7]
```diff
@@ -1103,6 +1181,140 @@ def normalize_args(**kwargs):
...
+    trace_fn: Callable[[Callable[..., Any], Iterable[Any]], torch.fx.GraphModule],
+    scalar_workaround,
+):
+    def get_file_template() -> str:
+        auto_generated_msg = textwrap.dedent(
+            """\
+            # This is an auto-generated file. Please do not modify it by hand.
+            # To re-generate, run:
+            # cd ~/pytorch && python
+            # torchgen/fuse_attention_patterns/gen_attention_patterns.py
```
[the remark this answers is not in the document]
Comment (aorenste) [reply torch/_inductor/pattern_matcher.py:1200]: The message is relevant - `gen_attention_patterns` will rebuild ALL the patterns (including the non-fuse ones).  If you want I can rename that script to something else but I worried that something/someone might be referencing that script so figured it would be easier to just have that script do all the work.

File: torch/_inductor/fx_passes/serialized_patterns/central_index.py [hunk 8]
```diff
@@ -1,123 +0,0 @@
...
-    '_sfdp_pattern_15_inference': _sfdp_pattern_15_inference,
-    '_sfdp_pattern_16_training': _sfdp_pattern_16_training,
-    '_sfdp_pattern_16_inference': _sfdp_pattern_16_inference,
-    '_sfdp_pattern_16_bs1_training': _sfdp_pattern_16_bs1_training,
-    '_sfdp_pattern_16_bs1_inference': _sfdp_pattern_16_bs1_inference,
-    '_sfdp_pattern_17_training': _sfdp_pattern_17_training,
-    '_sfdp_pattern_17_inference': _sfdp_pattern_17_inference,
-    '_sfdp_pattern_18_training': _sfdp_pattern_18_training,
-    '_sfdp_pattern_18_inference': _sfdp_pattern_18_inference,
-    '_sfdp_pattern_18_bs1_training': _sfdp_pattern_18_bs1_training,
```
[the remark this answers is not in the document]
Comment (aorenste) [reply torch/_inductor/fx_passes/serialized_patterns/central_index.py:65]: They should be - there's just no central index any more (which is what I thought you asked for).  As long as they're listed by _get_sfdp_patterns() (which was needed for them to be generated previously) then they should still be generated.

[the remark this answers is not in the document]
Comment (eellison) [reply torch/_inductor/fx_passes/serialized_patterns/central_index.py:65]: Right no central index is great. We still want a way of updating the patterns. The num_users were not changed, so I'm suspicious we are not generating them as part of this script anymore

[the remark this answers is not in the document]
Comment (aorenste) [reply torch/_inductor/fx_passes/serialized_patterns/central_index.py:65]: I'm not sure why `num_users` would change (OTOH - I also don't actually know what `num_users` tracks) because the only thing that changed was how we generate the scripts not any usage of those scripts. But the patterns are definitely being updated. If I manually change it and then regen then they're being re-written:
```
(base) $ hg diff
diff --git a/torch/_inductor/fx_passes/serialized_patterns/_sfdp_pattern_18.py b/torch/_inductor/fx_passes/serialized_patterns$
--- a/torch/_inductor/fx_passes/serialized_patterns/_sfdp_pattern_18.py
+++ b/torch/_inductor/fx_passes/serialized_patterns/_sfdp_pattern_18.py
@@ -205,6 +205,7 @@
 view_default_11 = CallFunction(aten.view.default, bmm_default_5, Ignored())
 permute_default_11 = CallFunction(aten.permute.default, view_default_11, Ignored())
 _sfdp_pattern_18_bs1_training = MultiOutputPattern([view_default_5,
+  "MANUALLY EDITED JUNK",
   permute_default_1,
   permute_default_3,
   permute_default_6,
(base) $ python torchgen/fuse_attention_patterns/gen_attention_patterns.py
(base) $ hg diff
(base) $
```

File: torch/_inductor/fx_passes/serialized_patterns/addmm_pattern.py (hunk 3 above)
Comment (eellison) [reply torch/_inductor/fx_passes/serialized_patterns/addmm_pattern.py:7]: Ah, I mistakenly thought it was only generating the attention patterns, and not yet the mm patterns. Maybe we can rename the file ?

**Review by eellison — APPROVED** [review]
Looks great - thanks for answering my questions!

**Comment by awgu** [comment]
I think because of this PR, I see the following when rebasing and trying to re-build PyTorch:
```
(pytorch-3.10) [andgu@devgpu011.cco1 /data/users/andgu/pytorch (unshard)]$ time TORCH_CUDA_ARCH_LIST="8.0 9.0" USE_CUDA=1 USE_DISTRIBUTED=1 python setup.py develop 2>&1 | tee build_log.txt
ninja: error: rebuilding 'build.ninja': '/data/users/andgu/pytorch/torchgen/fuse_attention_patterns/gen_attention_patterns.py', needed by 'aten/src/ATen/core_generated_declarations_yaml.cmake', missing and no known rule to make it
Building wheel torch-2.4.0a0+git288cae5
-- Building version 2.4.0a0+git288cae5
cmake --build . --target install --config Release -- -j 48
```

**Comment by aorenste** [comment]
> I think because of this PR, I see the following when rebasing and trying to re-build PyTorch:
> 
> ```
> (pytorch-3.10) [andgu@devgpu011.cco1 /data/users/andgu/pytorch (unshard)]$ time TORCH_CUDA_ARCH_LIST="8.0 9.0" USE_CUDA=1 USE_DISTRIBUTED=1 python setup.py develop 2>&1 | tee build_log.txt
> ninja: error: rebuilding 'build.ninja': '/data/users/andgu/pytorch/torchgen/fuse_attention_patterns/gen_attention_patterns.py', needed by 'aten/src/ATen/core_generated_declarations_yaml.cmake', missing and no known rule to make it
> Building wheel torch-2.4.0a0+git288cae5
> -- Building version 2.4.0a0+git288cae5
> cmake --build . --target install --config Release -- -j 48
> ```

Can you try adding "--cmake" after `develop`? That will kick the system to regenerate cmake files.

**Comment by awgu** [comment]
@aorenste Ah, thanks for the tip. I already ran `git clean` and am rebuilding now. Sorry for the noise!

**Comment by aorenste** [comment]
> @aorenste Ah, thanks for the tip. I already ran `git clean` and am rebuilding now. Sorry for the noise!

From my experience I generally follow this order until one of them works:
0. Run "setup.py develop"
1. Run "setup.py develop --cmake"
2. Run "setup.py clean && setup.py develop"
3. Run "git clean && setup.py develop"
4. Throw computer out window.

**Comment by xuzhao9** [comment]
This PR seems to cause sam_fast CUDA memory leak: https://github.com/pytorch/benchmark/issues/2231

Is this behavior expected?

**Comment by xuzhao9** [comment]
> > This PR seems to cause sam_fast CUDA memory leak on torchbench sam_fast model: [pytorch/benchmark#2231](https://github.com/pytorch/benchmark/issues/2231)
> > Is this behavior change expected?
> 
> No - not expected. I've been trying to repro it but having trouble getting the benchmark to run - I'll keep trying.

Created https://github.com/pytorch/pytorch/issues/124081, please reach out there if you have any problems reproducing, thanks!
